torch/csrc/jit/tensorexpr/cpp_codegen.cpp

```diff
@@ -1,44 +1,650 @@
+#include <algorithm>
+#include <type_traits>
+#include <vector>
+
 #include <torch/csrc/jit/tensorexpr/cpp_codegen.h>
+#include <torch/csrc/jit/tensorexpr/cpp_intrinsics.h>
+#include <torch/csrc/jit/tensorexpr/external_functions_registry.h>
+#include <torch/csrc/jit/tensorexpr/types.h>
 
 namespace torch {
 namespace jit {
 namespace tensorexpr {
 
-void CppPrinter::visit(const Allocate* alloc) {
-  constexpr size_t kAllocOnStackThresholdSize = 512;
+// Rewrites the variables' name according to valid C++ naming convention.
+// E.g. in Graph IR, variable name may contain '.', in C++, they are replaced
+// with '_'.
+class CppVarNameRewriter : public IRMutator {
+ public:
+  const Expr* mutate(const Var* v) override {
+    constexpr char kDot = '.';
+    constexpr char kUnderscore = '_';
+    if (v->name_hint().find(kDot) == std::string::npos) {
+      return v;
+    }
+    std::string name = v->name_hint();
+    std::replace(name.begin(), name.end(), kDot, kUnderscore);
```

Comment:
I think we could make `name_hint` in `Var` mutable. This way we would not need to create new variables when we do simple renaming and could get rid of the `getNewVar` method.

Reply:
If we make `name_hint` mutable, it simplifies this particular case. But I thought the design principle for `Expr` and `Stmt` is that their members are immutable, and if we want to mutate them, we should use `IRMutator`?

Reply:
For renamings, there is really no value to create a new `Var`, I think making `name_hint` mutable is fine.

Reply:
Yeah, I think it would be fine too. You are correct about Expr/Stmt mutability, but I think in this case it's better to mutate than clone everything. We're not changing the structure of exprs when we do this, so it should work fine I believe.